geomesa-gs-security/src/main/scala/org/geomesa/gs/security/audit/ElasticRequestDataListener.scala

```diff
@@ -0,0 +1,119 @@
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.http.HttpResponse
+import org.apache.http.client.config.RequestConfig
+import org.apache.http.client.methods.{HttpPost}
+import org.apache.http.concurrent.FutureCallback
+import org.apache.http.entity.StringEntity
+import org.apache.http.impl.nio.client.{CloseableHttpAsyncClient, HttpAsyncClients}
+import org.geoserver.monitor.{RequestData, RequestDataListener}
+import org.opengis.geometry.BoundingBox
+
+import scala.concurrent.Await
+
+
+class ElasticRequestDataListener extends RequestDataListener with LazyLogging {
+
+  private val requestConfig = RequestConfig.custom()
+    .setSocketTimeout(10000)
+    .setConnectTimeout(10000).build();
+
+  val client: CloseableHttpAsyncClient = HttpAsyncClients.custom()
+    .setMaxConnTotal(10)
+    .setDefaultRequestConfig(requestConfig)
+    .build()
+
+  client.start()
+  val host = sys.env.getOrElse("ELASTICSEARCH_HOST", null)
+  val index = sys.env.getOrElse("GEOSERVER_ES_INDEX", null)
+
+  private val gson: Gson = new GsonBuilder()
+    .registerTypeAdapter(classOf[Date], new DateSerializer)
+    .registerTypeAdapter(classOf[BoundingBox], new BoundingBoxSerializer)
+    .registerTypeAdapter(classOf[Throwable], new ThrowableSerializer)
+    .serializeNulls().create()
+
+  class DateSerializer extends JsonSerializer[Date] {
+    override def serialize(src: Date, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = {
+      new JsonPrimitive(src.getTime)
+    }
+  }
+
+   class BoundingBoxSerializer extends JsonSerializer[BoundingBox] {
+     override def serialize(src: BoundingBox, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = {
+       new JsonPrimitive(src.toString)
+     }
+   }
+
+  class ThrowableSerializer extends JsonSerializer[Throwable] {
+    override def serialize(src: Throwable, typeOfSrc: Type, context: JsonSerializationContext): JsonElement = {
+      new JsonPrimitive(src.getMessage)
+    }
+  }
+
+  override def requestStarted(requestData: RequestData): Unit = {}
+
+  override def requestUpdated(requestData: RequestData): Unit = {
+    writeToElasticsearch(requestData)
+  }
+
+  override def requestCompleted(requestData: RequestData): Unit = {
+    //writeToElasticsearch(requestData)
+  }
+
+  private def writeToElasticsearch(requestData: RequestData) = {
+    val post = new HttpPost(host + "/" + index + "/_doc?pretty")
+    post.addHeader("Content-Type", "application/json")
+
+    // 1. Skip over requests which do not have the resources set.
+    // 2. Skip over failures without the endTime set.
+    if (
+      !requestData.getResources.isEmpty &&
+      (!(requestData.getStatus == RequestData.Status.FAILED && requestData.getEndTime == null))
+    ) {
+      try{
+        val json = gson.toJson(requestData)
+        gson.fromJson(json, classOf[Any])
+        post.setEntity(new StringEntity(json))
+        val future = (client.execute(post, new LoggingCallback(post)))
```

Comment:
is there a reason we not using the elastic search high-level rest client? seems like it would simplify things here

Reply:
There is a lot to be done with dependencies, so the goal was to simplify that aspect by implementing it ourselves

Reply:
hmm, i see what you mean it does have a lot of dependencies. still, given the unexplained memory issues maybe it would be better to use it? @jnh5y what do you think? we could create an installer zip...

Reply:
ah, @jnh5y can probably show you some tools to dig into the memory usage.